oci/default.nix

```diff
@@ -0,0 +1,556 @@
+           };
+         };
+         hosts = true;
+         tmp = true;
+         usrBinEnv = "${pkgs.busybox}/bin/env";
+         binSh = "${pkgs.busybox}/bin/sh";
+       }
+
+     This will create an entries file setting up 2 users, root and nobody. It
+     also sets up hosts resolution, /tmp and links for /bin/sh and
+     /usr/bin/env.
+  */
+  makeFilesystem = {
+    accounts ? null,
+    hosts ? false,
+    tmp ? false,
+    usrBinEnv ? null,
+    binSh ? null,
+  }:
+  (lib.optionalAttrs (accounts != null) (
+    makeAccounts accounts
+  )) // (lib.optionalAttrs hosts {
+    "etc/nsswitch.conf" = {
+      type = "file";
+      text = "hosts: files dns";
+    };
+  }) // (lib.optionalAttrs tmp {
+    "tmp" = {
+      type = "directory";
+      mode = "0777";
+    };
+  }) // (lib.optionalAttrs (usrBinEnv != null) {
+    "usr/bin/env" = {
+      type = "link";
+      target = usrBinEnv;
+    };
+  }) // (lib.optionalAttrs (binSh != null) {
+    "bin/sh" = {
+      type = "link";
+      target = binSh;
+    };
+  });
+
+  makeNixStoreEntries = name: closure:
+    let
+      info = closureInfo { rootPaths = closure; };
+      store = runCommand "${name}-store" {} ''
+        export NIX_STATE_DIR=$out
+        ${nix}/bin/nix-store --load-db < ${info}/registration
+        rm $out/gcroots/profiles
+        rm $out/db/reserved
+      '';
+    in
+    {
+      "nix/var/nix" = {
+        type = "directory";
+        sources = [{
+          path = store;
+        }];
+      };
+    };
+
+  /* Wraps a derivation such that including it in a layer only includes its
+     dependencies
+  */
+  makeDependencyOnlyWrapper = drv:
+    let
+      referencesFile = writeReferencesToFile drv;
+    in
+    runCommand "${drv.name}-dependencies" {
+      passthru.excludeFromLayer = true;
+    } ''
+      grep -v "${drv}" "${referencesFile}" > $out
+    '';
+
+  /* Creates a layer from entries without its dependencies
+
+     This is mainly for internal use. See makeLayer for the structure of
+     entries.
+  */
+  makeBareLayerFromEntries = {
+    name,
+    entries,
+    umask
+  }: runCommand name {
+    entriesJson = builtins.toJSON entries;
+    passAsFile = [ "entriesJson" ];
+    passthru.excludeFromLayer = true;
+  } ''
+    mkdir -p $out
+    ${python3}/bin/python ${./build-layer.py} $out ${umask} "$SOURCE_DATE_EPOCH" < $entriesJsonPath
+    ${python3}/bin/python ${./hash-layer.py} $out
+  '';
+
+  /* Creates a layer from an optional base tar file and include/excludes
+
+     This is mainly for internal use. Prefer using makeLayer instead.
+  */
+  makeLayer' = {
+    name,
+    baseTar ? null,
+    includes ? [],
+    excludes ? [],
+    passthru ? {}
+  }:
+    let
+      includesFile = writeText "layer-includes" (builtins.concatStringsSep "\n" includes);
+      excludesFile = writeText "layer-excludes" (builtins.concatStringsSep "\n" excludes);
+      directExcludes = builtins.filter (x: x.excludeFromLayer or false) includes;
```

Comment:
`excludeFromLayer` is a custom attribute I'm adding to derivations whose dependencies we want but not actually include themselves. This is mostly useful for things like `writeReferencesToFile` or bare layers.